agent/src/action.rs

```diff
@@ -0,0 +1,156 @@
+        }
+
+        match start_exec(&exec_id).await {
+            Ok(StartExecResults::Attached {
+                mut output,
+                input: _,
+            }) => {
+                spawn(async move {
+                    while let Some(log) = output.next().await {
+                        match log {
+                            Ok(log_output) => {
+                                let _ = &log_input.lock().unwrap().send(Ok(ActionResponseStream {
+                                    log: log_output.to_string(),
+                                    action_id: action_id.lock().unwrap().to_string(),
+                                    result: Some(ActionResult {
+                                        completion: 2,
+                                        exit_code: None,
+                                    }),
+                                }));
+                            }
+                            Err(e) => {
+                                let _ = log_input.lock().unwrap().send(Ok(ActionResponseStream {
+                                    log: "Step exited with an error".to_string(),
+                                    action_id: action_id.lock().unwrap().to_string(),
+                                    result: Some(ActionResult {
+                                        completion: 3,
+                                        exit_code: Some(1),
+                                    }),
+                                }));
+                                return Err(Status::aborted(format!("Execution error: {}", e)));
+                            }
+                        }
+                    }
+                    Ok(())
+                });
+            }
+            Ok(StartExecResults::Detached) => {
+                return Err(Status::aborted("Can't attach to container"))
+            }
+            Err(_) => return Err(Status::aborted("Error happened when launching action")),
+        }
+
+        loop {
+            let exec_state = match inspect_exec(&exec_id).await {
+                Ok(exec_state) => exec_state,
+                Err(_) => return Err(Status::aborted("Error happened checking state of a step")),
+            };
+            match exec_state.exit_code {
+                Some(0) => {
+                    break;
+                }
+                Some(exit_code) => {
+                    info!("Step exited with an error: {}", exit_code);
+                    return Err(Status::aborted("Step exited with an error"));
+                }
+                None => {}
+            }
+            match exec_state.running {
+                Some(true) => {}
+                Some(false) => {
+                    break;
+                }
+                None => {
+                    return Err(Status::aborted("Error happened checking state of a step"));
+                }
+            }
+            sleep(Duration::from_secs(1)).await;
+        }
```

Comment:
(I think?) this could be broken down into another function, to improve clarity/readability? (less indents)

Reply:
Sound fair